File: sdk/dart/lib/Network.dart
import './NetworkTimestamp.dart';
import './CryptoTypes.dart';
import 'package:pointycastle/digests/ripemd160.dart';
import 'dart:typed_data';

class Network {
  String name;
  int identifier;
  NetworkTimestampDatetimeConverter datetimeConverter;
  Function addressHasher;
  Function createAddress;
  Type addressClass;
  NetworkTimestamp Function(int) networkTimestampFactory;

  Network(
    this.name, 
    this.identifier, 
    this.datetimeConverter,
    this.addressHasher,
    this.createAddress,
    this.addressClass,
    this.networkTimestampFactory,
  );
  
  Uint8List publicKeyToAddress(PublicKey publicKey) {
    var partOneHashBuilder = addressHasher();
    var partOneHash = partOneHashBuilder.process(publicKey.bytes);
    var partTwoHash = RIPEMD160Digest().process(partOneHash);
    var version = Uint8List.fromList([identifier, ...partTwoHash]);
    var partThreeHashBuilder = addressHasher();
    var checksum = partThreeHashBuilder.process(version).sublist(0, 4);
    return createAddress(version, checksum);
  }

  DateTime toDatetime(NetworkTimestamp referenceNetworkTimestamp) {
    return datetimeConverter.toDatetime(referenceNetworkTimestamp.timestamp);
  }

  NetworkTimestamp fromDatetime(DateTime referenceDatetime) {
    return networkTimestampFactory(datetimeConverter.toDifference(referenceDatetime));
  }
}

class NetworkLocator {
  static Network findByName(List<Network> networks, dynamic singleOrMultipleNames) {
    var names = singleOrMultipleNames is List<String> ? singleOrMultipleNames : [singleOrMultipleNames];
    try {
      return networks.firstWhere((network) => names.any((name) => name == network.name));
    } catch (e) {
      throw RangeError('no network found with name ${names.join(', ')}');
    }
  }

  static Network findByIdentifier(List<Network> networks, dynamic singleOrMultipleIdentifiers) {
    var identifiers = singleOrMultipleIdentifiers is List<int> ? singleOrMultipleIdentifiers : [singleOrMultipleIdentifiers];
    try {
      return networks.firstWhere((network) => identifiers.any((identifier) => identifier == network.identifier));
    } catch (e) {
      throw RangeError('no network found with identifier ${identifiers.join(', ')}');
    }
  }
}

File: sdk/php/src/Network.php
<?php

namespace SymbolSdk;

use RangeException;

class Network
{
  public string $name;
  public $identifier;
  public $datetimeConverter;
  private $_addressHasher;
  private $_createAddress;
  public $addressClass;
  public $networkTimestampClass;
  public function __construct($name, $identifier, $datetimeConverter, $addressHasher, $createAddress, $addressClass, $networkTimestampClass)
  {
    $this->name = $name;
    $this->identifier = $identifier;
    $this->datetimeConverter = $datetimeConverter;
    $this->_addressHasher  = $addressHasher;
    $this->_createAddress  = $createAddress;
    $this->addressClass = $addressClass;
    $this->networkTimestampClass = $networkTimestampClass;
  }

  /**
   * hash_update($hasher, pack('V', $nonce));
   * hash_update($hasher, $ownerAddress->binaryData);
   * $digest = hash_final($hasher, true);
   */
  public function publicKeyToAddress(string $publicKey)
  {
    $partOneHash = hash($this->_addressHasher, $publicKey, true);
    $partTwoHash = hash('ripemd160', $partOneHash, true);
    $identifierBinary = pack('C', $this->identifier);
    $version = $identifierBinary . $partTwoHash;
    $partThreeHash = hash($this->_addressHasher, $version, true);
    $checksum = substr($partThreeHash, 0, 4);
    return call_user_func($this->_createAddress, $version, $checksum);
  }
}

class NetworkLocator
{
  public static function findByName($networks, $singleOrMultipleNames)
  {
    $names = is_array($singleOrMultipleNames) ? $singleOrMultipleNames : [$singleOrMultipleNames];
    $matchingNetwork = null;

    foreach ($networks as $network) {
      if (in_array($network->name, $names)) {
        $matchingNetwork = $network;
        break;
      }
    }

    if (null === $matchingNetwork) {
      throw new RangeException("No network found with name '" . implode(', ', $names) . "'");
    }

    return $matchingNetwork;
  }

  public static function findByIdentifier($networks, $singleOrMultipleIdentifiers)
  {
    $identifiers = is_array($singleOrMultipleIdentifiers) ? $singleOrMultipleIdentifiers : [$singleOrMultipleIdentifiers];
    $matchingNetwork = null;

    foreach ($networks as $network) {
      if (in_array($network->identifier, $identifiers)) {
        $matchingNetwork = $network;
        break;
      }
    }

    if (null === $matchingNetwork) {
      throw new RangeException("No network found with name '" . implode(', ', $identifiers) . "'");
    }

    return $matchingNetwork;
  }
}
